lib/cdo/url_converter.rb

```diff
@@ -0,0 +1,47 @@
+class UrlConverter
+  LEARN_CODE_ORG_REGEX = /\/\/learn.code.org(?=$|\/)/
+  HOUROFCODE_COM_REGEX = /\/\/hourofcode\.com(?=$|\/)/
+  CSEDWEEK_ORG_REGEX = /\/\/csedweek\.org(?=$|\/)/
+  DASHBOARD_REGEX = /\/\/studio.code.org(?=$|\/)/
+
+  # For reference, a 'host' is a domain and (optionally) port without a protocol.
+  # Examples: code.org, studio.code.org, localhost-studio.code.org:3000
+  # @see https://developer.mozilla.org/en-US/docs/Web/API/Location
+  def initialize(dashboard_host: nil, pegasus_host: nil, hourofcode_host: nil,
+    csedweek_host: nil)
+    @dashboard_host = dashboard_host
+    @pegasus_host = pegasus_host
+    @hourofcode_host = hourofcode_host
+    @csedweek_host = csedweek_host
+  end
+
+  # An 'origin' is a protocol, domain, and (optional) port.  This method may
+  # replace all three.
+  def replace_origin(url)
+    if @dashboard_host
+      raise 'Should not use learn.code.org' unless LEARN_CODE_ORG_REGEX.match(url).nil?
+    end
+
+    if @hourofcode_host && HOUROFCODE_COM_REGEX =~ url
+      url = url.gsub(HOUROFCODE_COM_REGEX, "//" + @hourofcode_host)
+    elsif @csedweek_host && CSEDWEEK_ORG_REGEX =~ url
+      url = url.gsub(CSEDWEEK_ORG_REGEX, "//" + @csedweek_host)
+    elsif @dashboard_host && DASHBOARD_REGEX =~ url
+      url = url.gsub(DASHBOARD_REGEX, "//" + @dashboard_host)
+    elsif @pegasus_host
+      # Handle i18n subdomains
```

Comment:
nit: once we add `advocacy.code.org` soon, we will have a subdomain that's not actually i18n related.

Reply:
A couple possible approaches to this:
1) Advocacy is built on pegasus, correct?  So the code should still work, it's just the comment that's wrong?
2) Could handle advocacy.code.org earlier in the if-statement (like studio.code.org is handled)
3) Could have the i18n subdomain logic limited to two-character subdomains.

Reply:
Yes, just the comment here will a tiny bit wrong.  Other Pegasus code might have similar assumptions, but nothing to worry about in this change.